### PR TITLE
Add blog post draft for TBAA

### PR DESCRIPTION
**Pull request by kq-li**

### CODE REVIEW

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+"""
+[[extra.authors]]
+name = "Kenneth Li"
++++
+
+Alias analysis allows compilers to determine which pointers may (or must) 
+refer to the same memory location. This is most useful for the purpose of 
+instruction reordering; if a compiler knows that two memory instructions refer 
+to different memory locations, it can switch the order of the instructions. In 
+_Type-Based Alias Analysis_, Diwan, McKinley, and Moss propose and evaluate
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Perhaps include a link to the paper?

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+[[extra.authors]]
+name = "Kenneth Li"
++++
+
+Alias analysis allows compilers to determine which pointers may (or must) 
+refer to the same memory location. This is most useful for the purpose of 
+instruction reordering; if a compiler knows that two memory instructions refer 
+to different memory locations, it can switch the order of the instructions. In 
+_Type-Based Alias Analysis_, Diwan, McKinley, and Moss propose and evaluate 
+alias analyses for Modula-3 (a type-safe language with in heritance) that use
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: inheritance

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+_SMFieldTypeRefs_ augments the former with a single pass through the 
+instructions to pick up assignment information. The analyses all do a may-alias 
+analysis; all alias pairs calculated are possible aliases, but all other pairs 
+are confirmed to not be aliases.
+
+## Algorithms
+The _TypeDecl_ analysis relies on a simple predicate: two referenced memory 
+access paths `p` and `q` may be aliases if and only if there exists a type that 
+subtypes both the type of `p` and the type of `q`. If such a type did not 
+exist, for example, then if `p` and `q`were aliases, the object they refer to
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Missing space.

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+refines the previous: _TypeDecl_ uses only type/subtype compatibility, 
+_FieldTypeDecl_ additionally considers field names and other properties, and 
+_SMFieldTypeRefs_ augments the former with a single pass through the 
+instructions to pick up assignment information. The analyses all do a may-alias 
+analysis; all alias pairs calculated are possible aliases, but all other pairs 
+are confirmed to not be aliases.
+
+## Algorithms
+The _TypeDecl_ analysis relies on a simple predicate: two referenced memory 
+access paths `p` and `q` may be aliases if and only if there exists a type that
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: It might be worth briefly defining what an "access path" is? It's worth at least mentioning, for example, that we're assuming an OO language.

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+## Algorithms
+The _TypeDecl_ analysis relies on a simple predicate: two referenced memory 
+access paths `p` and `q` may be aliases if and only if there exists a type that 
+subtypes both the type of `p` and the type of `q`. If such a type did not 
+exist, for example, then if `p` and `q`were aliases, the object they refer to 
+would have two different, incompatible types. _FieldTypeDecl_ takes this 
+further by adding checks for field names, array accesses, and pointer 
+dereferences. For example, paths `p.f` and `q.g` may be aliases only if the 
+field names `f` and `g` are exactly the same and `p` and `q` themselves may be 
+aliases. Additionally, a pointer dereference `q^` may alias a field `p.f` or an
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: As above, it's probably a good idea to define that `q^` means in the assumed language (because most readers will not have programmed in Modula-3, but each concept has a direct analog in mainstream 2020 languages).

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+would have two different, incompatible types. _FieldTypeDecl_ takes this 
+further by adding checks for field names, array accesses, and pointer 
+dereferences. For example, paths `p.f` and `q.g` may be aliases only if the 
+field names `f` and `g` are exactly the same and `p` and `q` themselves may be 
+aliases. Additionally, a pointer dereference `q^` may alias a field `p.f` or an 
+array access `p[i]` only if its address was ever taken by the program (since 
+otherwise no pointer to it could exist). Further checks exist for comparisons 
+between pointers and array accesses, which the authors sum up as follows:
+
+1.	Identical [access paths] always alias each other.
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Is this a quote from the paper? If so, please indicate that it is (either with quotation marks or a blockquote). If not, I'm not sure what the square brackets indicate?

Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Also, it is always true that identical access paths alias? Or is it just that they always *may alias* (which is different)?

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+2.	Two qualified expressions may be aliases if they access the same field in
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: What is a "qualified expression"?

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+aliases. Additionally, a pointer dereference `q^` may alias a field `p.f` or an 
+array access `p[i]` only if its address was ever taken by the program (since 
+otherwise no pointer to it could exist). Further checks exist for comparisons 
+between pointers and array accesses, which the authors sum up as follows:
+
+1.	Identical [access paths] always alias each other.
+2.	Two qualified expressions may be aliases if they access the same field in 
+potentially the same object.
+3.	A pointer dereference may reference the same location as a qualified or 
+subscripted expression only if their types are compatible and the program may
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Similarly, what is a "subscripted expression"?

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+otherwise no pointer to it could exist). Further checks exist for comparisons 
+between pointers and array accesses, which the authors sum up as follows:
+
+1.	Identical [access paths] always alias each other.
+2.	Two qualified expressions may be aliases if they access the same field in 
+potentially the same object.
+3.	A pointer dereference may reference the same location as a qualified or 
+subscripted expression only if their types are compatible and the program may 
+take the address of the qualified or subscripted expression.
+4.	See (3)
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: What's this all about? (Why is this a separate list item?)

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+5.	In Modula-3, a subscripted expression cannot alias a qualified expression.
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Why? Is this relevant to the main idea here?

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+6.	Two subscripted expressions are aliases if they may subscript the same 
+array. _FieldTypeDecl_ ignores the actual subscripts.
+7.	For all other cases of [access paths], including two pointer dereferences,
+_FieldTypeDecl_ uses _TypeDecl_ to determine aliases.
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: I think this list needs some additional context-setting. It's not really clear what the reader is supposed to get out of reading this list of items, or if the order is significant. There is also lots of suprirrisng terminology that it kind of hard to parse without additional context.

Overall, I'm not sure you really need the list itself… is there any way to summarize what role this reasoning plays in the general idea? Doing so might at least make the list possible to understand, and it might even obviate the need for the list altogether.

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+The paper does a surprisingly thorough evaluation job considering the paper 
+was written in 1998. The authors observe that previous alias analyses were 
+evaluated by their static properties, like the size of the set of alias pairs, 
+and their dynamic properties, such as the impact of an alias analysis on a 
+compiler optimization that uses it. The static evaluation, performed over a 
+selection of 10 benchmarks, finds that _TypeDecl_ performs much worse than 
+_FieldTypeDecl_ in terms of alias set size, but _SMFieldTypeRefs_ barely 
+improves on _FieldTypeDecl_ at all. It also shows that applying type-based 
+alias analysis interprocedurally generates huge numbers of aliases, rendering 
+this type of analysis infeasible for global optimization.
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Do you mean interprocedural or global? (It can't be both.)

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+performance benefits, and don’t lend themselves to comparing two different 
+alias analyses – in both cases, it’s hard to tell whether the disambiguated 
+pointers will be relevant for the intended use case. Dynamic properties, on the 
+other hand, suffer from overspecificity – it’s difficult to evaluate general 
+efficacy of an alias analysis from performance improvements on a few specific 
+optimizations and benchmarks. Worst of all, neither evaluation can give an idea 
+of how much better an analysis could be.
+
+To alleviate these problems, the paper introduces the concept of limit 
+evaluation to figure out whether there were missed optimization opportunities
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: Maybe italicize "limit evaluation"? Also, FWIW, I think the more common term for this kind of thing today is "limit study."

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+pointers will be relevant for the intended use case. Dynamic properties, on the 
+other hand, suffer from overspecificity – it’s difficult to evaluate general 
+efficacy of an alias analysis from performance improvements on a few specific 
+optimizations and benchmarks. Worst of all, neither evaluation can give an idea 
+of how much better an analysis could be.
+
+To alleviate these problems, the paper introduces the concept of limit 
+evaluation to figure out whether there were missed optimization opportunities 
+due to undetected aliases. The authors instrumented their benchmarks to record 
+the address and value of every load, allowing them to realize exactly how many
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md]: realize -> measure/study/observe?

File: content/blog/2020-10-28-type-based-alias-analysis.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+The significance of this work is not to be understated; alias analyses over 
+types instead of instructions run much more efficiently due to the vastly 
+reduced search space, and the results in paper indicate that the precision 
+tradeoff could be minimal for some applications. From a modern perspective, 
+it seems that compiler developers have come to the same conclusion; many 
+modern compilers implement and support type-based alias analyses for many of 
+their optimization passes. Among these compilers are GCC and LLVM, showing that 
+the benefits have even extended outside the paper’s original realm of type-safe 
+languages.
```
Comment (sampsyo) [inline content/blog/2020-10-28-type-based-alias-analysis.md:128]: No need to add anything here, but I think it's interesting that the reason this stuff works in C/C++ is because the standards make it illegal to create two differently-typed references to the same value. Nothing actually stops programs from doing that and lots of software out there does it, of course, but all that software is broken, which is why TBAA is allowed.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Looks good overall! I am a bit confused about one section of the technical description, which may need a little rethinking and simplification.

File: content/blog/2020-10-28-type-based-alias-analysis.md (hunk 4 above)
Comment (kq-li) [reply content/blog/2020-10-28-type-based-alias-analysis.md]: clarified with analogy to C `*q`

File: content/blog/2020-10-28-type-based-alias-analysis.md (hunk 5 above)
Comment (kq-li) [reply content/blog/2020-10-28-type-based-alias-analysis.md]: clarified with analogy to C's `*q` syntax

File: content/blog/2020-10-28-type-based-alias-analysis.md (hunk 11 above)
Comment (kq-li) [reply content/blog/2020-10-28-type-based-alias-analysis.md]: good point, i removed the list and added a summary to indicate the authors' work in determining alias rules between different AP types

File: content/blog/2020-10-28-type-based-alias-analysis.md (hunk 12 above)
Comment (kq-li) [reply content/blog/2020-10-28-type-based-alias-analysis.md]: whoops, confused global with program wide

File: content/blog/2020-10-28-type-based-alias-analysis.md (hunk 13 above)
Comment (kq-li) [reply content/blog/2020-10-28-type-based-alias-analysis.md]: done! i kept "limit evaluation" because that's how they introduced it in the paper; notably, it's impossible to google for the terms "limit evaluation" or "limit study" without getting a bunch of math stuff 😛

File: content/blog/2020-10-28-type-based-alias-analysis.md (hunk 15 above)
Comment (kq-li) [reply content/blog/2020-10-28-type-based-alias-analysis.md:128]: ah! i was wondering what the breakthrough could have been for this to work outside type safety, but this seems a lot simpler, haha

**Comment by sampsyo** [comment]
Looking pretty good! When you're ready, use GitHub's "request review" option (in the upper-right).

**Review by sampsyo — APPROVED** [review]
Excellent!
